File: C++/CreationalDesignPattern/Abstract_FactoryDesign.cpp
#include <iostream>
#include <memory>
#include <stdexcept>

// Products in UI

struct IButton {
    virtual void renderButton() = 0;
    virtual ~IButton() = default;
};

struct IModal {
    virtual void renderModal() = 0;
    virtual ~IModal() = default;
};

struct IScreen {
    virtual void renderScreen() = 0;
    virtual ~IScreen() = default;
};

// Windows
struct WindowButton : IButton {
    void renderButton() override { std::cout << "Window button rendered" << std::endl; }
};

struct WindowModal : IModal {
    void renderModal() override { std::cout << "Window modal rendered" << std::endl; }
};

struct WindowScreen : IScreen {
    void renderScreen() override { std::cout << "Window screen rendered" << std::endl; }
};

// Mac
struct MacButton : IButton {
    void renderButton() override { std::cout << "Mac button rendered" << std::endl; }
};

struct MacModal : IModal {
    void renderModal() override { std::cout << "Mac modal rendered" << std::endl; }
};

struct MacScreen : IScreen {
    void renderScreen() override { std::cout << "Mac screen rendered" << std::endl; }
};

// Linux
struct LinuxButton : IButton {
    void renderButton() override { std::cout << "Mac button rendered" << std::endl; }
};

struct LinuxModal : IModal {
    void renderModal() override { std::cout << "Mac modal rendered" << std::endl; }
};

struct LinuxScreen : IScreen {
    void renderScreen() override { std::cout << "Mac screen rendered" << std::endl; }
};

struct IUIFactory {
    virtual std::unique_ptr<IButton> getButton() = 0;
    virtual std::unique_ptr<IModal> getModal() = 0;
    virtual std::unique_ptr<IScreen> getScreen() = 0;
    virtual ~IUIFactory() = default;
};

struct LinuxFactory : IUIFactory {
    std::unique_ptr<IButton> getButton() override { return std::make_unique<LinuxButton>(); }
    std::unique_ptr<IModal> getModal() override {
        throw std::runtime_error("Unimplemented method 'getModal'");
    }
    std::unique_ptr<IScreen> getScreen() override {
        throw std::runtime_error("Unimplemented method 'getScreen'");
    }
};

struct WindowUIFactory : IUIFactory {
    std::unique_ptr<IButton> getButton() override { return std::make_unique<WindowButton>(); }
    std::unique_ptr<IModal> getModal() override { return std::make_unique<WindowModal>(); }
    std::unique_ptr<IScreen> getScreen() override { return std::make_unique<WindowScreen>(); }
};

struct MacUIFactory : IUIFactory {
    std::unique_ptr<IButton> getButton() override { return std::make_unique<MacButton>(); }
    std::unique_ptr<IModal> getModal() override { return std::make_unique<MacModal>(); }
    std::unique_ptr<IScreen> getScreen() override { return std::make_unique<MacScreen>(); }
};

class UIRender {
public:
    std::unique_ptr<IButton> button;
    std::unique_ptr<IModal> modal;
    std::unique_ptr<IScreen> screen;

    explicit UIRender(IUIFactory& factory) {
        button = factory.getButton();
        modal = factory.getModal();
        screen = factory.getScreen();
        renderUI();
    }

    void renderUI() {
        button->renderButton();
        modal->renderModal();
        screen->renderScreen();
    }

    void toggleUI(IUIFactory& factory) {
        button = factory.getButton();
        modal = factory.getModal();
        screen = factory.getScreen();
        renderUI();
    }
};

int main() {
    WindowUIFactory windowFactory;
    UIRender myView(windowFactory);

    MacUIFactory macFactory;
    myView.toggleUI(macFactory);

    return 0;
}

File: Python/CreationalDesignPattern/Abstract_FactoryDesign.py
from abc import ABC, abstractmethod


# Products in UI

class IButton(ABC):
    @abstractmethod
    def render_button(self):
        pass


class IModal(ABC):
    @abstractmethod
    def render_modal(self):
        pass


class IScreen(ABC):
    @abstractmethod
    def render_screen(self):
        pass


# Windows
class WindowButton(IButton):
    def render_button(self):
        print("Window button rendered")


class WindowModal(IModal):
    def render_modal(self):
        print("Window modal rendered")


class WindowScreen(IScreen):
    def render_screen(self):
        print("Window screen rendered")


# Mac
class MacButton(IButton):
    def render_button(self):
        print("Mac button rendered")


class MacModal(IModal):
    def render_modal(self):
        print("Mac modal rendered")


class MacScreen(IScreen):
    def render_screen(self):
        print("Mac screen rendered")


# Linux
class LinuxButton(IButton):
    def render_button(self):
        print("Mac button rendered")


class LinuxModal(IModal):
    def render_modal(self):
        print("Mac modal rendered")


class LinuxScreen(IScreen):
    def render_screen(self):
        print("Mac screen rendered")


class IUIFactory(ABC):
    @abstractmethod
    def get_button(self) -> IButton:
        pass

    @abstractmethod
    def get_modal(self) -> IModal:
        pass

    @abstractmethod
    def get_screen(self) -> IScreen:
        pass


class LinuxFactory(IUIFactory):
    def get_button(self) -> IButton:
        return LinuxButton()

    def get_modal(self) -> IModal:
        raise NotImplementedError("Unimplemented method 'get_modal'")

    def get_screen(self) -> IScreen:
        raise NotImplementedError("Unimplemented method 'get_screen'")


class WindowUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return WindowButton()

    def get_modal(self) -> IModal:
        return WindowModal()

    def get_screen(self) -> IScreen:
        return WindowScreen()


class MacUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return MacButton()

    def get_modal(self) -> IModal:
        return MacModal()

    def get_screen(self) -> IScreen:
        return MacScreen()


class UIRender:
    def __init__(self, factory: IUIFactory):
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()

    def render_ui(self):
        self.button.render_button()
        self.modal.render_modal()
        self.screen.render_screen()

    def toggle_ui(self, factory: IUIFactory):
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()


def main():
    my_view = UIRender(WindowUIFactory())
    my_view.toggle_ui(MacUIFactory())


if __name__ == "__main__":
    main()
